File: 1027-longest-arithmetic-subsequence/1027-longest-arithmetic-subsequence.cpp
class Solution {
public:
    int longestArithSeqLength(vector<int>& nums) {
        
        vector<unordered_map<int, int>> dp(nums.size());
        int diff;
        int res = 0;

        for (int i = 0; i < nums.size(); i++) {
            for (int j = 0; j < i; j++) {
                diff = nums[i] - nums[j];
                dp[i][diff] = 1 + (dp[j].contains(diff) ? dp[j].at(diff) : 1);
            }
            for (auto pair : dp[i]) {
                res = max(res, pair.second);
            }
        }

        return res;
    }
};

File: 1027-longest-arithmetic-subsequence/1027-longest-arithmetic-subsequence.py
from functools import cache

class Solution:
    def longestArithSeqLength(self, nums: List[int]) -> int:

        dp = [{} for i in range(0, len(nums))]
        res = 0
        for i in range(0, len(nums)):
            for j in range(0, i):
                dp[i][nums[i] - nums[j]] = 1 + dp[j].get(nums[i] - nums[j], 1)
            for v in dp[i].values():
                res = max(res, v)
        
        return res
